### PR TITLE
8260630: Templatize literal_size

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review a couple of small cleanups in Hashtable support.

(1) Replace some "default" overloads for literal_size with a template.  This
makes it easier to add new Hashtable value types whose sizes aren't
interesting for statistics printout.

(2) Trivial change to remove no longer used instantiation of Hashtable
support for oop values.

Testing:
mach5 tier1

/issue add 8260629

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8260630](https://bugs.openjdk.java.net/browse/JDK-8260630): Templatize literal_size
 * [JDK-8260629](https://bugs.openjdk.java.net/browse/JDK-8260629): Remove explicit instantiation of Hashtable with oop value

### Reviewers
 * [Per Liden](https://openjdk.java.net/census#pliden) (@pliden - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2310/head:pull/2310`
`$ git checkout pull/2310`

### CODE REVIEW

**Review by pliden — APPROVED** [review]
Looks good to me!

**Review by tschatzl — APPROVED** [review]
Lgtm.

File: src/hotspot/share/utilities/hashtable.cpp [hunk 1]
```diff
@@ -235,10 +231,6 @@ template <class T, MEMFLAGS F> TableStatistics Hashtable<T, F>::statistics_calcu
 }
 
 // Dump footprint and bucket length statistics
-//
-// Note: if you create a new subclass of Hashtable<MyNewType, F>, you will need to
-// add a new function static int literal_size(MyNewType lit)
-// because I can't get template <class T> int literal_size(T) to pick the specializations for Symbol and oop.
```
Comment (coleenp) [inline src/hotspot/share/utilities/hashtable.cpp:241]: I wrote that.  Thanks for fixing this!

File: src/hotspot/share/utilities/hashtable.cpp [hunk 2]
```diff
@@ -110,22 +110,18 @@ template <MEMFLAGS F> void BasicHashtable<F>::free_buckets() {
...
-static int literal_size(Klass*)        { return 0; }
-static int literal_size(nmethod*)      { return 0; }
+// Default overload, for types that are uninteresting.
+template<typename T> static int literal_size(T) { return 0; }
 
 static int literal_size(Symbol *symbol) {
   return symbol->size() * HeapWordSize;
 }
 
 static int literal_size(oop obj) {
```
Comment (coleenp) [inline src/hotspot/share/utilities/hashtable.cpp:120]: Is literal_size(oop) still needed?

**Review by coleenp — APPROVED** [review]
Looks good!

File: src/hotspot/share/utilities/hashtable.cpp (hunk 2 above)
Comment (kimbarrett) [reply src/hotspot/share/utilities/hashtable.cpp:120]: It's used by literal_size(WeakHandle).

**Comment by kimbarrett** [comment]
Thanks @pliden , @coleenp , @tschatzl for reviews.

**Pull request closed without merging**